oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/profile/analyze/ProfileStackNode.java

```diff
@@ -0,0 +1,242 @@
+        emptyNode.children = new LinkedList<>();
+        return emptyNode;
+    }
+
+    /**
+     * accumulate {@link ProfileStack} to this tree, it will invoke on the tree root node
+     * @param stack
+     */
+    public void accumulateFrom(ProfileStack stack) {
+        List<String> stackList = stack.getStack();
+        if (codeSignature == null) {
+            codeSignature = stackList.get(0);
+        }
+        // increase self
+        this.increase(stack);
+
+        // handle stack children
+        ProfileStackNode parent = this;
+        for (int childDepth = 1; childDepth < stackList.size(); childDepth++) {
+            String elementCodeSign = stackList.get(childDepth);
+
+            // find same code signature children
+            ProfileStackNode childElement = null;
+            for (ProfileStackNode child : parent.children) {
+                if (Objects.equal(child.codeSignature, elementCodeSign)) {
+                    childElement = child;
+                    break;
+                }
+            }
+
+            if (childElement != null) {
+                // increase child node
+                childElement.increase(stack);
+                parent = childElement;
+            } else {
+                // add children
+                ProfileStackNode childNode = createEmptyNode();
+                childNode.codeSignature = elementCodeSign;
+                childNode.increase(stack);
+
+                parent.children.add(childNode);
+                parent = childNode;
+            }
+        }
+    }
+
+    /**
+     * combine from other {@link ProfileStackNode}
+     * @param node
+     * @return
+     */
+    public ProfileStackNode combine(ProfileStackNode node) {
+        // combine this node
+        this.combineThisNode(node);
+
+        // merge tree using pre order
+        LinkedList<Pair<ProfileStackNode, List<ProfileStackNode>>> stack = new LinkedList<>();
+        stack.add(new Pair<>(this, node.children));
+        while (!stack.isEmpty()) {
+            Pair<ProfileStackNode, List<ProfileStackNode>> needCombineNode = stack.pop();
+            ProfileStackNode combineTo = needCombineNode.key;
+
+            // merge node
+            for (ProfileStackNode needCombineChildren : needCombineNode.value) {
+                boolean sameReference = false;
+                // find node
+                ProfileStackNode combinedNode = combineTo.children.stream().filter(n -> needCombineChildren.matches(n)).findFirst().orElse(null);
+
+                if (combinedNode == null) {
+                    combineTo.children.add(needCombineChildren);
+                    combinedNode = needCombineChildren;
+                } else if (needCombineChildren != combinedNode) {
+                    combinedNode.combineThisNode(needCombineChildren);
+                } else {
+                    // if same reference, dont need to combine anymore
+                    sameReference = true;
```

Comment:
Why same reference happens?

Reply:
It's a bug, has fixed. If combine node from B -> A, A node not found, just need to add children node, don't need to keep merge child.